### ISSUE TITLE
[USERSNAP] Approval / Slashing votes bars show all votes

### ISSUE THREAD

**Issue by dmtrjsg** [Joystream/pioneer#2715]
**Taken at**: https://dao.joystream.org/#/proposals/preview/20
**Fields**:
- Houston, we have a problem.: Approval / Slashing votes bars show all votes
- Give us some details: There are [obviously](https://discord.com/channels/811216481340751934/812343711870091285/958287033514725377) different opinions but in my mind the blue bar signals how many specific approve / slash votes have been cast, not the total number of votes.
- Blue bar indicates total number of votes.
- Discord Handle: l1dev

![SCREENSHOT](https://resources.usersnap.com/company/f9f719df-a929-4461-bb6e-46dbfbceffb7/datapoint_screenshot/ef61bed8-9019-4de7-8eea-29133edddc5c-annotated_77866e05-4876-457c-9886-238c513e3e20.png?etag=acbe64065ed7a8fb724b1292fc89eb3b)
*[Download original image](https://resources.usersnap.com/company/f9f719df-a929-4461-bb6e-46dbfbceffb7/datapoint_screenshot/ef61bed8-9019-4de7-8eea-29133edddc5c-annotated_77866e05-4876-457c-9886-238c513e3e20.png?etag=acbe64065ed7a8fb724b1292fc89eb3b)*

**Browser**: Chrome 99 (Linux)
**Screen size**: 1920x1080
**Browser size**: 2390x1216
**Labels**: bug-non-critical
**[Open #160 in Usersnap](https://app.usersnap.com/l/feedback/551c0630-70c2-4977-8f40-900ac5a2b7d3?utm_medium=integration&utm_campaign=integration&utm_content=open)**

Powered by **[Usersnap](https://usersnap.com/?utm_source=product&amp;utm_medium=poweredbylink&amp;utm_campaign=github_entry)**.

**Comment by traumschule** [comment]
![votes-counter](https://user-images.githubusercontent.com/31551045/160591621-7be86729-5098-41d5-ad6c-85217336b8c3.png)

**Comment by mkbeefcake** [comment]
@isonar Can you please assign this issue for me?

**Comment by singulart** [comment]
Assigned to @mkbeefcake

**Comment by thesan** [comment]
Sorry for being late on this one.
The current behaviour is not accounting for abstentions votes, but apart from that it is mostly correct.

TLDR: IMO the scope should be:
- [ ] Make the quorum `total - abstain` instead of just `total`
- [ ] Add `abstain` to [the story](https://pioneer-2-storybook.vercel.app/?path=/story/proposals-proposalpreview-proposalstatistics--default)
- [ ] Maybe improve the design

### The quorum on chain

As it is written [here](https://github.com/Joystream/joystream/blob/fb9049a9652dc91329c8db972fbf85b7c8b7e48b/runtime-modules/proposals/engine/src/types/mod.rs#L309) the Quorum is the votes count without abstentions, then this gets divided by the council size (total potential votes), and:
- For [approval](https://github.com/Joystream/joystream/blob/fb9049a9652dc91329c8db972fbf85b7c8b7e48b/runtime-modules/proposals/engine/src/types/mod.rs#L309) it is compared to the `approvalQuorumPercentage`.
- For [slashing](https://github.com/Joystream/joystream/blob/fb9049a9652dc91329c8db972fbf85b7c8b7e48b/runtime-modules/proposals/engine/src/types/mod.rs#L357) it is compared to the `slashingQuorumPercentage`.

### Here is a concrete example

Lets take the signal proposal (from https://polkadot.js.org/apps/#/chainstate/constants):
```
const proposalsCodex.signalProposalParameters: ProposalParameters
{
  votingPeriod: 72,000
  gracePeriod: 0
  approvalQuorumPercentage: 60
  approvalThresholdPercentage: 80
  slashingQuorumPercentage: 60
  slashingThresholdPercentage: 80
  requiredStake: 25,000
  constitutionality: 1
}
```
The `approvalQuorumPercentage` is 60%. So for a council size of 5 the approval quorum 3.

If 2 councilors vote to approve this proposal [the bars](https://pioneer-2-storybook.vercel.app/?path=/story/proposals-proposalpreview-proposalstatistics--default) currently look like this:
![image](https://user-images.githubusercontent.com/6571453/188854662-9f26eee7-8ae7-489a-9b58-de68aca9a2fc.png)
So the threshold is reached but the quorum is not.

If another councilor cast an "Abstain" vote (we currently can't represent this on the story), the Approval Quorum bar would increase and reach the tick mark, while in fact if the deciding period were to end in this state the proposal would get rejected due to the Approval Quorum not being.

Instead if a councilor rejects the proposal:
![image](https://user-images.githubusercontent.com/6571453/188855756-f32eafd6-ad20-4e5d-8745-737f72822beb.png)
Now the approval quorum is reached (i.e enough councilor have expressed their opinions for the approval threshold to matter) but the approval threshold is no longer reached.

### UX issues

- The UX doesn't make it clear that rejection just means not approved and not slashed i.e one of: approval quorum, slashing quorum, approval threshold, or slashing threshold was not reached. Maybe indicating that the proposal is e.g "Rejected so far" would help.
- I can't remember why there is a "max" on the slashing side e.g "max 3 votes" and "max 80%". Personally it confuses me as it imply that slashing relies on a fundamentally different logic than approval, which is not the case.

**Comment by mkbeefcake** [comment]
Hello @thesan 

do you have discord id? i hope to discuss this via chat more.

**Comment by singulart** [comment]
@thesan an extensive piece of research in here ☝🏻 please give a reply whenever you got a minute.

**Comment by thesan** [comment]
> 1. can you let me know your discord id? I hope to discuss via discord to save the time.

I just contacted you there.

> 2. I didn't figure out the reject values inside Rust code, so didn't use it in this frontend side.

As I mentioned earlier:
> rejection just means not approved and not slashed i.e one of: approval quorum, slashing quorum, approval threshold, or slashing threshold was not reached

You can see this in 2 places in the Rust code:
Here for early rejection:
![image](https://user-images.githubusercontent.com/6571453/189326962-e1211a84-5c4d-453a-a596-528e435e9795.png)
And here it's the last fallback when the vote ends 
![image](https://user-images.githubusercontent.com/6571453/189327216-8f445108-98b2-421c-9d12-8a25147b55b1.png)

> 3. saying about 'rejection' UX, do you have any idea for it? 

Kind of but it's not very clear yet so it shouldn't be implemented in the PR. FYI the idea would be to show the "virtual state of the proposal". Meaning it starts rejected but if both threshold and quorum are reach for approval it becomes approved (but it should be clear that it's just a virtual state not the actual state of the proposal). It would require some design and I need to talk to @dmtrjsg about it. So it's definitely not to be implemented now!

> 4. add abstain to story  : i can guess that is related with ProposalStatistics.stories.tsx, but i am not sure how to check it in my local end.

The command to run storybook is: `yarn storybook`. As I mentioned in the PR review, it takes a while and throws a bunch of warning but it should start eventually. Also it automatically opens itself on your default browser once it's up, so you don't need to check on it all the time.

**Comment by mkbeefcake** [comment]
Hello @thesan 
I tried to update the source code, please check it again

**Comment by dmtrjsg** [comment]
Thanks all for looking into this issue and bringing the rust conditions up to the ticket. Makes it easy to take it forward.

I think the best would be to drive it by design team, so we solve this key part once and for good.

Few thoughts of the top of my head as I believe we should:
- Add abstention bar as this is a key part of calculating the thresholds
- Add explainer that unless both thresholds are met, the proposal will be rejected by the time council period ends
- Add a note that threshold is dynamic and depends on abstentions count and formulas for calculation of thereof to the tooltips
- Also I noticed that slashing quorum bar increments with additional votes count, this should also be fixed in the subsequent PR.

also @mkbeefcake we will review your latest PR with @thesan and take a view whether it makes sense to merge it for incremental benefit before reviewing this widget in a more holistic way involving the design team.

**Comment by ivanturlakov** [comment]
@dmtrjsg pls let me know the best endpoint to test it since this PR merged in staging where we have only one working env https://34.235.154.53.nip.io/network/config.json with one council seat

**Comment by dmtrjsg** [comment]
@ivanturlakov its fine, we can roll it out and double check on prod later, as there were so many reviews of the code already and it looks solid and the story book holds up, and they are using the same code. 👍
